lib/pbench/server/database/models/users.py

```diff
@@ -0,0 +1,144 @@
+        return {
+            "username": self.username,
+            "email": self.email,
+            "first_name": self.first_name,
+            "last_name": self.last_name,
+            "registered_on": self.registered_on,
+        }
+
+    @staticmethod
+    def query(id=None, username=None, email=None):
+        # Currently we would only query with single argument. Argument need to be either username/id/email
+        try:
+            if username:
+                user = (
+                    Database.db_session.query(User).filter_by(username=username).first()
+                )
+            elif id:
+                user = Database.db_session.query(User).filter_by(id=id).first()
+            elif email:
+                user = Database.db_session.query(User).filter_by(email=email).first()
+            else:
+                user = None
+        except Exception as e:
+            raise e
+        else:
+            return user
+
+    def add(self):
+        """
+        Add the current user object to the database
+        """
+        try:
+            Database.db_session.add(self)
+            Database.db_session.commit()
+        except Exception as e:
+            Database.db_session.rollback()
+            raise e
+
+    # Prevent update on "registered_on" field
+    @validates("registered_on")
+    def evaluate_registered_on(self, key, value):
+        if self.registered_on:  # Field already exists
+            if self.registered_on.strftime("%Y-%m-%d %H:%M:%S.%f") != value:
+                raise ValueError("registered_on cannot be modified.")
+        return value
+
+    # validate the email field
+    @validates("email")
+    def evaluate_email(self, key, value):
+        try:
+            valid = validate_email(value)
+            email = valid.email
+        except Exception as e:
+            raise e
+        else:
+            return email
+
+    def update(self, **kwargs):
+        """
+        Update the current user object with given keyword arguments
+        """
+        try:
+            for key, value in kwargs.items():
+                if key == "auth_tokens":
+                    # Insert the auth token
+                    self.auth_tokens.append(value)
+                    Database.db_session.add(value)
+                elif key == "password":
+                    setattr(self, key, generate_password_hash(value))
+                else:
+                    setattr(self, key, value)
+            Database.db_session.commit()
+        except Exception as e:
+            Database.db_session.rollback()
+            raise e
+
+    @staticmethod
+    def delete(username):
+        """
+        Delete the user with a given username except admin
+        :param username:
+        :return:
+        """
+        try:
+            user_query = Database.db_session.query(User).filter_by(username=username)
+            # Do not delete if the user is admin
+            if not user_query.first().is_admin():
```

Comment:
It would be good, once we have the notion, to ensure that there is at least one existing user with admin privileges.  We presumably want to allow users to delete themselves, and we presumably do not want to allow users to casually delete other users; however, admin users should be allowed to delete other users, but, for safety sake, admin users should not be allowed to delete themselves.  So, the test here devolves to:
- if the target of the deletion is not the current user, then the current user must be an admin;
- if the target of the deletion is the current user, then the current user must not be an admin;
- otherwise, proceed.

(You can do this nicely with an XOR operator: `if username == current ^ current.is_admin(): ...`. )

Reply:
And, there's another wrinkle:  I'm a fan of separating policy from mechanism...so, perhaps this function should just delete any user that it is told to, and the burden should be placed on the _caller_ of this function to avoid undesirable deletions.

Among other things, this approach frees this module of the responsibility for knowing whether a given user is an admin or not.  (That is, is this the right module for housing that interface?)

Reply:
The original idea here was that, at least initially, there will be a reserved "admin" user defined with privileges that can't be granted to any other account.

It would, however, be cleaner to clearly separate the concepts of "has administrative access" and "is the administrator account" in the API now rather than deal with problems later, so you're not wrong.

Reply:
The `User` model interface right now only has a check for `is_admin` it does not distinguish between  "has administrative access" and "is the administrator account". I think it will be easier to manage this interface when we have role based authentication and a notion of groups and have restriction policies defined for a group. Created an issue https://github.com/distributed-system-analysis/pbench/issues/2126